server/src/test/java/org/solarbank/server/integration/EnergySavingControllerIT.java

```diff
@@ -0,0 +1,74 @@
+
+import com.fasterxml.jackson.core.JsonProcessingException;
+import com.fasterxml.jackson.databind.ObjectMapper;
+import org.junit.jupiter.api.Test;
+import org.solarbank.server.ErrorMessage;
+import org.solarbank.server.ValidationMessage;
+import org.solarbank.server.dto.CalculateRequest;
+import org.springframework.http.MediaType;
+import static org.solarbank.server.CreateMockCalculateRequest.createCalculateRequest;
+import static org.springframework.test.web.servlet.request.MockMvcRequestBuilders.post;
+import static org.springframework.test.web.servlet.result.MockMvcResultHandlers.print;
+import static org.springframework.test.web.servlet.result.MockMvcResultMatchers.jsonPath;
+import static org.springframework.test.web.servlet.result.MockMvcResultMatchers.status;
+
+public class EnergySavingControllerIT extends IntegrationTestBase {
+
+    private final ObjectMapper requestMapper = new ObjectMapper();
+
+    private String toStringCalculateRequest(CalculateRequest calculateRequest)
+            throws JsonProcessingException {
+        return requestMapper.writeValueAsString(calculateRequest);
+    }
+
+    @Test
+    public void energySavingController_validRequest_status200() throws Exception {
+        CalculateRequest calculateRequest = createCalculateRequest();
+
+        String validRequest = toStringCalculateRequest(calculateRequest);
+
+        mockMvc.perform(post("/v1.0/api/calculate")
+                        .contentType(MediaType.APPLICATION_JSON)
+                        .content(validRequest))
+                .andExpect(status().isOk())
+                .andDo(print())
+                .andExpect(jsonPath("$.EnergyGenPerYear").value(1.0))
+                .andExpect(jsonPath("$.EnergyGenPerMonth.January").value(0.1))
+                .andExpect(jsonPath("$.EnergyGenPerMonth.February").value(0.2))
+                .andExpect(jsonPath("$.SavingsPerYear.CurrencyCode").value("USD"))
+                .andExpect(jsonPath("$.SavingsPerYear.Amount").value(1000.0));
+    }
+
+    @Test
+    public void energySavingController_invalidRequest_status400() throws Exception {
+        CalculateRequest calculateRequest = createCalculateRequest();
+
+        calculateRequest.getLocation().setLatitude(190.0);
+
+        String invalidRequest = toStringCalculateRequest(calculateRequest);
+
+        mockMvc.perform(post("/v1.0/api/calculate")
+                        .contentType(MediaType.APPLICATION_JSON)
+                        .content(invalidRequest))
+                .andExpect(status().isBadRequest())
+                .andDo(print())
+                .andExpect(jsonPath("$.Error.Code").value(400))
+                .andExpect(jsonPath("$.Error.Status").value(ErrorMessage.BAD_REQUEST.getMessage()))
+                .andExpect(jsonPath("$.Error.Message").value(ValidationMessage.LATITUDE_MAX.getMessage() + "; "));
+    }
+
+    @Test
+    public void energySavingController_emptyRequest_status500() throws Exception {
```

Comment:
This is an interesting test as it highlights an issue with our code. If the request is empty then this should be handled by our validation code, and we should not return an internal server error to the client via an unhandled exception. Instead we should be returning a 400 as it's a client error.

Can we validate the request body is not null/ empty with a javax validation, and add a test here to cover that case?

Reply:
added a test for null request body. But had to add an exception handler for HttpMessageNotReadableException taht is thrown by spring when request body is null